compiler/rustc_query_system/src/dep_graph/graph.rs

```diff
@@ -288,29 +297,42 @@ impl<K: DepKind> DepGraph<K> {
             let task_deps = Lock::new(TaskDeps::default());
             let result = K::with_deps(Some(&task_deps), op);
             let task_deps = task_deps.into_inner();
+            let task_deps = task_deps.reads;
 
-            // The dep node indices are hashed here instead of hashing the dep nodes of the
-            // dependencies. These indices may refer to different nodes per session, but this isn't
-            // a problem here because we that ensure the final dep node hash is per session only by
-            // combining it with the per session random number `anon_id_seed`. This hash only need
-            // to map the dependencies to a single value on a per session basis.
-            let mut hasher = StableHasher::new();
-            task_deps.reads.hash(&mut hasher);
-
-            let target_dep_node = DepNode {
-                kind: dep_kind,
-                // Fingerprint::combine() is faster than sending Fingerprint
-                // through the StableHasher (at least as long as StableHasher
-                // is so slow).
-                hash: data.current.anon_id_seed.combine(hasher.finish()).into(),
-            };
+            let dep_node_index = match task_deps.len() {
+                0 => {
+                    // Dependency-less anonymous nodes can safely be replaced by a dummy node.
```

Comment:
Let's provide an explanation here how this actually works. Something like:
```
// Because the dep-node id of anon nodes is computed from the sets of its dependencies 
// we already know what the ID of this dependency-less node is going to be (i.e. equal 
// to the precomputed `SINGLETON_DEPENDENCYLESS_ANON_NODE`). As a consequence we can 
// skip creating a `StableHasher` and sending the node through interning.